src/Analysis/Ast/Impl/Analyzer/Symbols/ClassEvaluator.cs

```diff
@@ -53,19 +55,50 @@ internal sealed class ClassEvaluator : MemberEvaluator {
                 // Set bases to the class.
                 var bases = new List<IPythonType>();
                 foreach (var a in _classDef.Bases.Where(a => string.IsNullOrEmpty(a.Name))) {
-                    // We cheat slightly and treat base classes as annotations.
-                    var b = Eval.GetTypeFromAnnotation(a.Expression);
-                    if (b != null) {
-                        var t = b.GetPythonType();
-                        bases.Add(t);
-                        t.AddReference(Eval.GetLocationOfName(a.Expression));
+                    // can either be the name of a class declared in the current module
+                    // or a class declared in some other module
+                    var cur = a.Expression as NameExpression;
+                    var mod = a.Expression as MemberExpression;
+                    if (cur == null && mod == null) {
+                        Eval.ReportDiagnostics(
+                                Eval.Module.Uri,
+                                new DiagnosticsEntry(
+                                    Resources.InheritNonClass.FormatInvariant(a.ToCodeString(Eval.Ast, CodeFormattingOptions.Traditional)),
+                                    _classDef.NameExpression.GetLocation(Eval.Module)?.Span ?? default,
+                                    Diagnostics.ErrorCodes.InheritNonClass,
+                                    Severity.Error,
+                                    DiagnosticSource.Analysis));
+                    }
+
+                    if (cur != null) {
+                        var arg = Eval.GetInScope(cur.Name, outerScope);
+                        if (arg is PythonClassType classType) {
+                            bases.Add(classType);
+                            // trying to extend a constant
+                        } else if (arg is IPythonConstant m) {
+                            Eval.ReportDiagnostics(
+                                Eval.Module.Uri,
+                                new DiagnosticsEntry(
+                                    Resources.InheritNonClass.FormatInvariant(m.Value),
+                                    _classDef.NameExpression.GetLocation(Eval.Module)?.Span ?? default,
```

Comment:
These 4 lines are similar to the previous case. You can extract it into a separate method.

Reply:
Yup!